counsel.el

```diff
@@ -5087,6 +5088,30 @@ this if you want to provide specific non-standard build types to
 `counsel-compile'.  The default helpers are setup to handle common
 build environments.")
 
+;; Should this be hook-able to deal with project managers and other
+;; version control systems?
+(defun counsel--find-project-root ()
+  "Traverse up the tree and find the project root."
+  (expand-file-name (locate-dominating-file
+                     (or (buffer-file-name) default-directory)
+                     ".git")))
+
+;; No easy way to make directory local, would buffer local make more sense?
```

Comment:
Remember `.dir-locals.el`?

Reply:
@Alexander-Shukaev sure you can create a .dir-locals which could seed compile history but how can you make a variable not in .dir-locals  be "dir local" to a subtree of the filesystem? If there was a make-variable-subtree-local then I would certainly use that over this hack.

Reply:
Well, yeah there are only `buffer-local` variables.  If your concern is maintaining one true version of history per project root directory in all of the buffers, then having a global `alist`, where each entry is a cells of project root directory and a history list itself looks like the only solution to me.

Reply:
Symbolic links also have to be taken into account in all of that.

Reply:
@Alexander-Shukaev Yes I want a version history per project root. Does the ivy framework have support for that? Would I need to overload the setting of the history variable myself? It would certainly be cleaner.

Reply:
I'm not aware of such support.  The solution has to be tailored here I guess.  How about the one I proposed?  I'm sure `projectile` does something similar to maintain configuration settings per project, could look for inspiration there.